### PR TITLE
Openssl 3.0.0+quic

### PR DESCRIPTION
**Pull request by tmshort**
Rebase onto 3.0. Very clean! Only CHANGES.md and VERSION.dat had conflects.

##### Checklist

- [X] documentation is added or updated
- [X] tests are added or updated

### CODE REVIEW

File: include/openssl/evp.h [hunk 1]
```diff
@@ -1739,6 +1739,7 @@ int EVP_PKEY_CTX_set_mac_key(EVP_PKEY_CTX *ctx, const unsigned char *key,
  * Method handles all operations: don't assume any digest related defaults.
  */
 # define EVP_PKEY_FLAG_SIGCTX_CUSTOM     4
+
```
Comment (nibanks) [inline include/openssl/evp.h:1742]: nit: necessary change?

File: ssl/statem/statem_clnt.c [hunk 2]
```diff
@@ -904,6 +904,14 @@ int ossl_statem_client_construct_message(SSL *s, WPACKET *pkt,
         break;
 
     case TLS_ST_CW_END_OF_EARLY_DATA:
+#ifndef OPENSSL_NO_QUIC
+        /* QUIC does not send EndOfEarlyData, draft-ietf-quic-tls-24 S8.3 */
```
Comment (nibanks) [inline ssl/statem/statem_clnt.c:908]: nit: any comments around drafts should probably reflect RFC status now.

File: ssl/statem/statem_quic.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (nibanks) [inline ssl/statem/statem_quic.c:2]: This copyright seems different from the other ones that say 2018-2021. Why?

**Review by nibanks — APPROVED** [review]
LGTM. A couple nits, but aren't really new here.

File: include/openssl/evp.h (hunk 1 above)
Comment (tmshort) [reply include/openssl/evp.h:1742]: Nope... we can clean it up on the next rebase (or a subsequent clean-up commit). It was probably left over from when I reverted any FIPS-checksum-impacting changes.

File: ssl/statem/statem_clnt.c (hunk 2 above)
Comment (tmshort) [reply ssl/statem/statem_clnt.c:908]: Yup, we can do that on a clean up commit

File: ssl/statem/statem_quic.c (hunk 3 above)
Comment (tmshort) [reply ssl/statem/statem_quic.c:2]: It's when the code was originally written! :D The copyright can be updated in a clean up commit.

**Comment by tmshort** [comment]
Checks are failing in "check_docs", specifically `make md-nits`; in at least one file that isn't modified by QUIC.
Another issue is in "external_test_pyca" test=`TestRSA.test_generate_rsa_keys[3-2048]`. It appears related to RSA key generation, which is not touched by QUIC.

I made a tweak to CHANGES.md, and I'm not seeing the same issue locally with `./config && make md-nits`. So...?

**Comment by kaduk** [comment]
> I made a tweak to CHANGES.md, and I'm not seeing the same issue locally with `./config && make md-nits`. So...?

I think it's "known" that md-nits will pick up different things depending on the version of the markdown libraries/tools you have installed

**Review by kaduk — APPROVED** [review]
Nice clean rebase.
Nick's comments are valid comments (as are the replies).
I will try to look through the changes themselves (not just the diff-of-diffs) asynchronously.

**Comment by tmshort** [comment]
Waiting on queued checks to complete before merging.

**Comment by tmshort** [comment]
In case it takes too long for GitHub to run them...
The original run: https://github.com/quictls/openssl/actions/runs/1209540222

File: ssl/statem/statem_clnt.c (hunk 2 above)
Comment (kaduk) [reply ssl/statem/statem_clnt.c:908]: (The man pages also have some draft references, too.)

**Comment by tmshort** [comment]
Added fixes for the nits and the "make update" issue.

**Comment by tmshort** [comment]
**CIFuzz / Fuzzing (pull_request)** seems to be failing due to something unrelated to QUIC:
```
: error: unsupported option '--with-fuzzer-lib=/usr/lib/libFuzzingEngine'

**Comment by tmshort** [comment]
Everything but the **CIFuzz** passed. @nibanks @kaduk

**Review by nibanks — APPROVED** [review]
Thanks for fixing my nits!

**Review by kaduk — APPROVED** [review]
Updates look good, thanks

**Pull request merged**